### PR TITLE
Duplicate SERP metrics in P3A

### LINKED ISSUE #58117: Add metrics for SERP queries per day
Add the following daily metrics:

How many Brave queries did you make in the past day?
`Brave.Search.SERPBrave`

0. 0
1. 1
2. 2
3. 3
4. 4
5. 5
6. 6
7. 7
8. 8
9. 9-12
10. 13-17
11. 18+

How many Google queries did you make in the past day?
`Brave.Search.SERPGoogle`

1. 1
2. 2
3. 3
4. 4
5. 5
6. 6
7. 7
8. 8
9. 9-13
10. 14-21
11. 22-34
12. 35+

How many non-Brave/Google queries did you make in the past day?
`Brave.Search.SERPOther`

1. 1
2. 2
3. 3
4. 4
5. 5+

How many search queries did you make between 1 - 28 days ago?
`Brave.Search.SERPStale`

1. 1
2. 2
3. 3
4. 4
5. 5
6. 6-10
7. 11-15
8. 16+

### PR DESCRIPTION
**Pull request by DJAndries**
- Resolves https://github.com/brave/brave-browser/issues/58117

### CODE REVIEW

File: browser/serp_metrics/serp_metrics_all_profiles_aggregator.h [hunk 1]
```diff
@@ -34,8 +36,11 @@ class SerpMetricsAllProfilesAggregator {
 
   virtual ~SerpMetricsAllProfilesAggregator();
 
-  virtual size_t GetSearchCountForYesterday(SerpMetricType type) const;
-  virtual size_t GetSearchCountForStalePeriod() const;
+  virtual size_t GetSearchCountForYesterday(
+      SerpMetricType type,
+      std::optional<base::Time> last_report_time = std::nullopt) const;
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_all_profiles_aggregator.h]: default args are banned for virtual functions: https://google.github.io/styleguide/cppguide.html#Default_Arguments

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+void SerpMetricsP3A::OnMetricCycled(const std::string& histogram_name) {
+  const std::string_view* dict_key =
+      base::FindOrNull(kSerpHistogramToDictKey, histogram_name);
+  if (!dict_key) {
+    return;
+  }
+  ScopedDictPrefUpdate update(local_state_.get(),
+                              prefs::kP3ALastReportedAtDict);
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_p3a.cc:120]: Maybe we should update `prefs::kP3ALastReportedAtDict` when we get metrics, i.e. when we call GetSearchCountForYesterday?

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  rotation_subscription_ = p3a_service->RegisterRotationCallback(
+      base::BindRepeating(&SerpMetricsP3A::OnRotation, base::Unretained(this)));
+  metric_cycled_subscription_ =
+      p3a_service->RegisterMetricCycledCallback(base::BindRepeating(
+          &SerpMetricsP3A::OnMetricCycled, base::Unretained(this)));
+}
+
+// static
+void SerpMetricsP3A::RegisterPrefs(PrefRegistrySimple* registry) {
+  registry->RegisterDictionaryPref(prefs::kP3ALastReportedAtDict);
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_p3a.cc:103]: can this be a time pref?

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+SerpMetricsP3A::~SerpMetricsP3A() = default;
+
+void SerpMetricsP3A::Init() {
+  if (!base::FeatureList::IsEnabled(kSerpMetricsFeature) ||
+      !kSerpMetricsP3A.Get()) {
+    return;
+  }
+
+  p3a::P3AService* p3a_service = g_brave_browser_process->p3a_service();
```
Comment (netzenbot) [inline browser/serp_metrics/serp_metrics_p3a.cc]: `SerpMetricsP3A` reaches for `g_brave_browser_process->p3a_service()` here, and for `g_browser_process->profile_manager()` in `ReportMetrics()`, rather than taking them as dependencies. `BraveBrowserProcessImpl::StartBraveServices()` (the only caller of `Init()`) already has both, so passing them in — e.g. `Init(p3a::P3AService&, ProfileAttributesStorage&)` — would also let you drop the `CHECK_IS_TEST()` branch, which today means the unit test bails out of `Init()` and never covers the callback-registration path at all. ([✅ Inject Dependencies at Construction; Don't Reach for Global State or `Browser*`](https://github.com/brave/brave-core/tree/master/docs/best-practices/architecture.md#ARCH-073))

File: browser/serp_metrics/BUILD.gn [hunk 5]
```diff
@@ -33,9 +35,13 @@ static_library("serp_metrics") {
   deps = [
     "//brave/browser:browser_process",
     "//brave/components/constants",
+    "//brave/components/p3a",
```
Comment (netzenbot) [inline browser/serp_metrics/BUILD.gn]: `serp_metrics_p3a.h` is listed in `public` and includes `brave/components/p3a/metric_log_type.h`, so `//brave/components/p3a` belongs in `public_deps` rather than `deps` — consumers of this target (e.g. `misc_metrics_impl`, which includes `serp_metrics_p3a.h` and has no `//brave/components/p3a` dep) otherwise don't inherit it. The other header-visible deps in this target (`//base`, `//brave/components/serp_metrics`, ...) are already in `public_deps`. ([✅ Use `public_deps` for Header-File Includes in BUILD.gn](https://github.com/brave/brave-core/tree/master/docs/best-practices/build-system.md#BS-024))

File: browser/serp_metrics/serp_metrics_p3a_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    TestingProfile* profile = profile_manager_->CreateTestingProfile("profile");
+
+    serp_metrics_ = std::make_unique<SerpMetrics>(
+        &local_state_,
+        ProfileAttributesTimePeriodStoreFactory(
+            profile->GetPath(), g_browser_process->profile_manager()
+                                    ->GetProfileAttributesStorage()));
+
+    serp_metrics_p3a_ = std::make_unique<SerpMetricsP3A>(local_state_);
+    serp_metrics_p3a_->Init();
```
Comment (netzenbot) [inline browser/serp_metrics/serp_metrics_p3a_unittest.cc]: `TestingBraveBrowserProcess::p3a_service()` returns nullptr (test/base/testing_brave_browser_process.cc:132), so this `Init()` always takes the `CHECK_IS_TEST()` early return — neither subscription is ever registered, and the `kSerpMetricsP3A` param gate is never exercised. `NothingRecordedUntilRotation` therefore passes trivially. Worth either injecting a `P3AService` so the wiring is covered, or dropping the `Init()` call so the tests don't imply coverage they don't have. ([✅ Verify Tests Actually Test What They Claim](https://github.com/brave/brave-core/tree/master/docs/best-practices/testing-isolation.md#TI-012))

Comment (DJAndries) [reply browser/serp_metrics/serp_metrics_p3a_unittest.cc]: Yes that's true today, but not necessarily in the future. What if the logic changes and it unintentionally reports a histogram without an `OnRotation` call?

File: browser/serp_metrics/serp_metrics_p3a.cc (hunk 3 above)
Comment (DJAndries) [reply browser/serp_metrics/serp_metrics_p3a.cc:103]: Each metric/count is sent separately, so using a single time pref instead of one time value  per metric could lead to inaccuracies

File: browser/serp_metrics/serp_metrics_p3a.cc (hunk 2 above)
Comment (DJAndries) [reply browser/serp_metrics/serp_metrics_p3a.cc:120]: to clarify, `OnMetricCycled` is called by the P3A service when the metric is transmitted, which matches the logic of the usage ping SERP metrics. If we updated the time value in `ReportMetrics`, we would be doing so prematurely

File: browser/serp_metrics/serp_metrics_all_profiles_aggregator.h (hunk 1 above)
Comment (DJAndries) [reply browser/serp_metrics/serp_metrics_all_profiles_aggregator.h]: thanks, i didn't know

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+}  // namespace
+
+SerpMetricsP3A::SerpMetricsP3A(PrefService& local_state)
+    : local_state_(local_state) {}
+
+SerpMetricsP3A::~SerpMetricsP3A() = default;
+
+void SerpMetricsP3A::Init(p3a::P3AService* p3a_service,
+                          ProfileManager* profile_manager) {
+  if (profile_manager) {
```
Comment (netzenbot) [inline browser/serp_metrics/serp_metrics_p3a.cc:79]: `profile_manager` is never null at the only call site (`BraveBrowserProcessImpl::StartBraveServices()`), and a null one would just move the failure to the `CHECK(profile_attributes_storage_)` in `ReportMetrics()`. Either take it by reference or mirror the `p3a_service` handling below with an explicit `CHECK_IS_TEST()`, so it's clear null is test-only. ([✅ Use `CHECK_IS_TEST` for Null Checks That Should Only Occur in Tests](https://github.com/brave/brave-core/tree/master/docs/best-practices/architecture.md#ARCH-024))

File: browser/serp_metrics/serp_metrics_p3a.h [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  SerpMetricsP3A& operator=(const SerpMetricsP3A&) = delete;
+
+  static void RegisterPrefs(PrefRegistrySimple* registry);
+
+  // Registers P3A rotation and metric-cycled callbacks. Must be called on the
+  // UI thread after `P3AService` is initialized.
+  void Init(p3a::P3AService* p3a_service, ProfileManager* profile_manager);
+
+  // Invoked via P3AService rotation callback.
+  void OnRotation(p3a::MetricLogType log_type);
```
Comment (netzenbot) [inline browser/serp_metrics/serp_metrics_p3a.h:51]: nit: `OnRotation()` and `OnMetricCycled()` are bound inside `Init()` and the only external callers are `serp_metrics_p3a_unittest.cc`, so they could be private with a `friend class SerpMetricsP3ATest;`. ([✅ Use Friend Class for Test/Private Access](https://github.com/brave/brave-core/tree/master/docs/best-practices/architecture.md#ARCH-020))

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 9]
```diff
@@ -0,0 +1,157 @@
+/* Copyright (c) 2026 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/serp_metrics/serp_metrics_p3a.h"
+
+#include <climits>
+#include <string>
+#include <string_view>
```
Comment (netzenbot) [inline browser/serp_metrics/serp_metrics_p3a.cc:10]: nit: `ReportMetric` uses `std::size_t` (template parameter) and `std::size(buckets)`, but this new file includes neither `<cstddef>` nor `<iterator>` — both come in transitively today. ([✅ Always Include What You Use (IWYU)](https://github.com/brave/brave-core/tree/master/docs/best-practices/coding-standards.md#CS-001))

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+  rotation_subscription_ = p3a_service->RegisterRotationCallback(
+      base::BindRepeating(&SerpMetricsP3A::OnRotation, base::Unretained(this)));
+  metric_cycled_subscription_ =
+      p3a_service->RegisterMetricCycledCallback(base::BindRepeating(
+          &SerpMetricsP3A::OnMetricCycled, base::Unretained(this)));
+}
+
+// static
+void SerpMetricsP3A::RegisterPrefs(PrefRegistrySimple* registry) {
```
Comment (netzenbot) [inline browser/serp_metrics/serp_metrics_p3a.cc:102]: nit: `RegisterPrefs()` is declared before `Init()` in the header but defined after it here; swapping the two definitions keeps the `.cc` order matching the `.h`. ([✅ Function Ordering in `.cc` Should Match `.h`](https://github.com/brave/brave-core/tree/master/docs/best-practices/coding-standards.md#CS-017))

File: components/p3a/metric_names.h [hunk 11]
```diff
@@ -319,6 +319,26 @@ inline constexpr auto kCollectedExpressHistograms =
       .ephemeral = true,
       .attributes = MetricAttributes{MetricAttribute::kIsBrowserDefault, MetricAttribute::kAnswerIndex, MetricAttribute::kCountryCode, MetricAttribute::kPlatform, MetricAttribute::kYoi},
     }},
+    {"Brave.Search.SERPBrave", MetricConfig{
```
Comment (netzenbot) [inline components/p3a/metric_names.h:382]: This registers four brand-new P3A metrics (`Brave.Search.SERPBrave`/`SERPGoogle`/`SERPOther`/`SERPStale`). Please post a brief description of them in the p3a Slack channel for privacy review before merging. ([✅ New P3A Metrics Require Privacy Review](https://github.com/brave/brave-core/tree/master/docs/best-practices/architecture.md#ARCH-068))

File: components/serp_metrics/pref_names.h [hunk 12]
```diff
@@ -19,6 +19,13 @@ inline constexpr std::string_view kDeprecatedSerpMetricsTimePeriodStorage =
 inline constexpr std::string_view kLastReportedAt =
     "brave.serp_metrics.last_reported_at";
 
+// Updated by `SerpMetricsP3A` whenever a P3A SERP metric is cycled (i.e. its
+// value has been sent). Each metric has its own last-reported time stored as a
+// dict entry keyed by the histogram name. Used as the stale-window cutoff for
```
Comment (netzenbot) [inline components/serp_metrics/pref_names.h:24]: nit: this comment doesn't match the code — `SerpMetricsP3A::OnMetricCycled()` keys the dict by the short keys (`brave`/`google`/`other`/`stale`) via `kSerpHistogramToDictKey`, not by histogram name, and `GetLastReportedTime()` is used as the cutoff for all four SERP metrics, not just `SERPStale`. ([✅ Delete Dead Documentation](https://github.com/brave/brave-core/tree/master/docs/best-practices/documentation.md#DOC-002))

File: components/p3a/metric_names.h (hunk 11 above)
Comment (DJAndries) [reply components/p3a/metric_names.h:382]: way ahead of you

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+  if (!dict_key) {
+    return;
+  }
+  ScopedDictPrefUpdate update(local_state_.get(),
+                              prefs::kP3ALastReportedAtDict);
+  update->Set(*dict_key, base::TimeToValue(base::Time::Now()));
+}
+
+void SerpMetricsP3A::ReportMetrics() {
+  CHECK(profile_attributes_storage_);
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_p3a.cc:125]: can we store `profile_attributes_storage_` as a raw_ref and pass `ProfileAttributesStorage` as a reference to Init?

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+}  // namespace
+
+SerpMetricsP3A::SerpMetricsP3A(PrefService& local_state)
+    : local_state_(local_state) {}
+
+SerpMetricsP3A::~SerpMetricsP3A() = default;
+
+void SerpMetricsP3A::Init(p3a::P3AService* p3a_service,
+                          ProfileManager* profile_manager) {
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_p3a.cc:78]: Seems like `ProfileManager* profile_manager` is only used to get  ProfileAttributesStorage. Maybe we can just pass ref to ProfileAttributesStorage and so extra dependency from ProfileManager can be removed

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+               /*suppress_when_zero=*/true);
+  ReportMetric(kStaleSerpHistogramName, kStaleSerpBuckets, stale_count,
+               /*suppress_when_zero=*/true);
+}
+
+base::Time SerpMetricsP3A::GetLastReportedTime(
+    std::string_view dict_key) const {
+  const base::Value* value =
+      local_state_->GetDict(prefs::kP3ALastReportedAtDict).Find(dict_key);
+  return value ? base::ValueToTime(*value).value_or(base::Time())
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_p3a.cc:153]: maybe this be simplified to first checking if (!value) {...}
if so we can just do simple return base::ValueToTime(*value).value_or(base::Time()) at the end.

File: browser/serp_metrics/serp_metrics_all_profiles_aggregator_unittest.cc [hunk 16]
```diff
@@ -108,9 +113,12 @@ TEST_F(SerpMetricsAllProfilesAggregatorTest,
 
   SerpMetricsAllProfilesAggregator aggregator(local_state(),
                                               profile_attributes_storage());
-  EXPECT_EQ(1U, aggregator.GetSearchCountForYesterday(SerpMetricType::kBrave));
-  EXPECT_EQ(1U, aggregator.GetSearchCountForYesterday(SerpMetricType::kGoogle));
-  EXPECT_EQ(1U, aggregator.GetSearchCountForYesterday(SerpMetricType::kOther));
+  EXPECT_EQ(1U, aggregator.GetSearchCountForYesterday(SerpMetricType::kBrave,
+                                                      std::nullopt));
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_all_profiles_aggregator_unittest.cc:117]: lets add arguments comments per google style guide:
https://google.github.io/styleguide/cppguide.html#Function_Argument_Comments

File: browser/serp_metrics/serp_metrics_all_profiles_aggregator_unittest.cc [hunk 17]
```diff
@@ -78,10 +80,13 @@ TEST_F(SerpMetricsAllProfilesAggregatorTest,
 
   SerpMetricsAllProfilesAggregator aggregator(local_state(),
                                               profile_attributes_storage());
-  EXPECT_EQ(0U, aggregator.GetSearchCountForYesterday(SerpMetricType::kBrave));
-  EXPECT_EQ(0U, aggregator.GetSearchCountForYesterday(SerpMetricType::kGoogle));
-  EXPECT_EQ(0U, aggregator.GetSearchCountForYesterday(SerpMetricType::kOther));
-  EXPECT_EQ(0U, aggregator.GetSearchCountForStalePeriod());
+  EXPECT_EQ(0U, aggregator.GetSearchCountForYesterday(SerpMetricType::kBrave,
+                                                      std::nullopt));
```
Comment (aseren) [inline browser/serp_metrics/serp_metrics_all_profiles_aggregator_unittest.cc:84]: We are only passing `std::nullopt` in tests here. Should we add tests that uses non-null time?

File: browser/serp_metrics/serp_metrics_p3a.cc (hunk 15 above)
Comment (DJAndries) [reply browser/serp_metrics/serp_metrics_p3a.cc:156]: i'm not sure if this would meaningfully simplify things

File: browser/serp_metrics/serp_metrics_p3a.cc (hunk 13 above)
Comment (DJAndries) [reply browser/serp_metrics/serp_metrics_p3a.cc:128]: `Init` gets called after construction, so `profile_attributes_storage_` is null until then, making raw_ref a bad fit

**Review by aseren — APPROVED** [review]

**Review by Olipro — APPROVED** [review]

File: browser/brave_stats/brave_stats_updater_params.cc [hunk 18]
```diff
@@ -198,24 +199,25 @@ GURL BraveStatsUpdaterParams::GetUpdateURL(
         update_url, "braveSearch",
         base::NumberToString(
             serp_metrics_aggregator->GetSearchCountForYesterday(
-                serp_metrics::SerpMetricType::kBrave)));
+                serp_metrics::SerpMetricType::kBrave, std::nullopt)));
```
Comment (tmancey) [inline browser/brave_stats/brave_stats_updater_params.cc]: param comments may help with these: `/*last_report_time=*/`

File: browser/serp_metrics/serp_metrics_all_profiles_aggregator.cc [hunk 19]
```diff
@@ -28,21 +28,27 @@ SerpMetricsAllProfilesAggregator::SerpMetricsAllProfilesAggregator(
 SerpMetricsAllProfilesAggregator::~SerpMetricsAllProfilesAggregator() = default;
 
 size_t SerpMetricsAllProfilesAggregator::GetSearchCountForYesterday(
-    SerpMetricType type) const {
+    SerpMetricType type,
+    std::optional<base::Time> last_report_time) const {
```
Comment (tmancey) [inline browser/serp_metrics/serp_metrics_all_profiles_aggregator.cc]: nit: just for consistency in rest of SERP, `last_reported_at`

File: browser/serp_metrics/serp_metrics_p3a.cc [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+#include "components/prefs/scoped_user_pref_update.h"
+
+namespace serp_metrics {
+
+namespace {
+
+constexpr char kBraveSerpDictKey[] = "brave";
+constexpr char kGoogleSerpDictKey[] = "google";
+constexpr char kOtherSerpDictKey[] = "other";
+constexpr char kStaleSerpDictKey[] = "stale";
```
Comment (tmancey) [inline browser/serp_metrics/serp_metrics_p3a.cc]: maybe `const std::string_view`

File: components/serp_metrics/serp_metrics.cc [hunk 21]
```diff
@@ -162,7 +166,15 @@ size_t SerpMetrics::GetSearchCountForTesting(SerpMetricType type) const {
 
 ///////////////////////////////////////////////////////////////////////////////
 
-base::Time SerpMetrics::GetStartOfStalePeriod() const {
+base::Time SerpMetrics::GetStartOfStalePeriod(
+    std::optional<base::Time> last_report_time) const {
+  if (last_report_time.has_value()) {
+    // Passed by `SerpMetricsP3A` from `kP3ALastReportedAtDict`. Null means
+    // the metric has never been cycled, so the full retention window is stale.
+    return last_report_time->is_null() ? base::Time()
```
Comment (tmancey) [inline components/serp_metrics/serp_metrics.cc]: `return last_report_time().value_or(base::Time())`

File: components/serp_metrics/serp_metrics.cc [hunk 22]
```diff
@@ -162,7 +166,15 @@ size_t SerpMetrics::GetSearchCountForTesting(SerpMetricType type) const {
 
 ///////////////////////////////////////////////////////////////////////////////
 
-base::Time SerpMetrics::GetStartOfStalePeriod() const {
+base::Time SerpMetrics::GetStartOfStalePeriod(
+    std::optional<base::Time> last_report_time) const {
+  if (last_report_time.has_value()) {
```
Comment (tmancey) [inline components/serp_metrics/serp_metrics.cc]: nit: `if (last_report_time) {`

**Review by tmancey — APPROVED** [review]

File: components/serp_metrics/serp_metrics.cc (hunk 21 above)
Comment (DJAndries) [reply components/serp_metrics/serp_metrics.cc]: we still need to call `UTCMidnight` if a non-null time value is passed

Comment (tmancey) [reply components/serp_metrics/serp_metrics.cc]: Ah thanks.
